app/mailers/user_mailer.rb

```diff
@@ -14,71 +17,87 @@ class UserMailer < ActionMailer::Base
     ),
   )
 
-  def email_confirmation_instructions(user, email, token, request_id:, instructions:)
-    with_user_locale(user) do
-      presenter = ConfirmationEmailPresenter.new(user, view_context)
+  def validate_user_and_email_address
+    @user = params.fetch(:user)
+    @email_address = params.fetch(:email_address)
+    if @user.id != @email_address.user_id
+      raise UserEmailAddressMisMatchError.new(
+        "User ID #{@user.id} does not match EmailAddress ID #{@email_address.id}",
+      )
+    end
+  end
+
+  def email_confirmation_instructions(token, request_id:, instructions:)
+    with_user_locale(@user) do
+      presenter = ConfirmationEmailPresenter.new(@user, view_context)
       @first_sentence = instructions || presenter.first_sentence
       @confirmation_period = presenter.confirmation_period
       @request_id = request_id
       @locale = locale_url_param
       @token = token
-      mail(to: email, subject: t('user_mailer.email_confirmation_instructions.subject'))
+      mail(
+        to: @email_address.email,
+        subject: t('user_mailer.email_confirmation_instructions.subject'),
+      )
     end
   end
 
-  def unconfirmed_email_instructions(user, email, token, request_id:, instructions:)
-    with_user_locale(user) do
-      presenter = ConfirmationEmailPresenter.new(user, view_context)
+  def unconfirmed_email_instructions(token, request_id:, instructions:)
+    with_user_locale(@user) do
```

Comment:
Could we avoid the diffs of `user` to `@user` by making it an `attr_reader` ?

Reply:
Good point, thank you. Added in d313a207c